ТЗ 8 : Добавлена работа с временем, тесты,

AntonyukV516 opened the pull request:
добавлена новая функциональность

AntonyukV516 committed:
ТЗ 8 : Добавлена работа с временем, тесты,
 добавлена проверка на пересечения

AntonyukV516 committed:
исправлен codeStyle

vlolad commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -98,22 +100,25 @@ public void deleteSubtaskById(Integer id) {
             for (Epic epic : epics) {
-                String epicAsString = String.format("%d,EPIC,%s,%s,%s,%d\n", epic.getId(), epic.getName(),
-                        epic.getStatus(), epic.getDescription(), -1);
+                String epicAsString = String.format("%d,EPIC,%s,%s,%s,%s,%d,%s,%d\n", epic.getId(), epic.getName(),
```
Форматирование задач лучше вынести в отдельный класс-форматтер со статическими методами на преобразование в строку и из неё. Так получится убрать лишнее дублирование и проще потом добавлять другие способы форматирования

Это по желанию, но очень рекомендую
Если не будешь делать отдельный класс - надо вынести строку формата в константу на уровне класса. Типы задач также в константы, после чего их добавлять наряду с другими параметрами формата

vlolad replied:
Ну и можно будет использовать StringBuilder/StringJoiner для того, чтобы собирать строку для сохранения)

vlolad commented on src/manager/InMemoryTaskManager.java:
```diff
@@ -222,4 +240,46 @@ public ArrayList<Subtask> getSubTaskFromEpic(Integer epicId) {
     }
+
+    @Override
+    public List<PreTask> getPrioritizedTasks() {
```
Не подойдет такая реализация. Надо сразу собирать список задач по приоритету сразу при добавлении, чтобы сортировка осуществлялась сразу же при их добавлении. Для этого надо объявить на уровне класса новый список prioritizedTasks с подходящей структурой для хранения (подойдет TreeSet). Сортировка по startTime (в конструкторе TreeSet потребуется указать  Comparator.comparing(Task::getStartTime)).

Таким образом в этом методе можно будет возвращать задачи за О(1), т.к. они уже отсортированы

vlolad commented on src/manager/InMemoryTaskManager.java:
```diff
@@ -222,4 +240,46 @@ public ArrayList<Subtask> getSubTaskFromEpic(Integer epicId) {
+    }
+
+    @Override
+    public boolean isIntersection(PreTask preTask1) {
```
Потребуется переработать метод с учетом комментария выше, надо оптимизировать

vlolad commented on src/manager/InMemoryTaskManager.java:
```diff
@@ -222,4 +240,46 @@ public ArrayList<Subtask> getSubTaskFromEpic(Integer epicId) {
+        if (startTime1 == null || endTime1 == null || startTime2 == null || endTime2 == null) {
+            return false;
+        }
+        return startTime1.isBefore(endTime2) && endTime1.isAfter(startTime2)
```
Близко к правде, но надо проверить все возможные случаи пересечений (включая те, когда новая задача начинается раньше и заканчивается позднее существующей, и наоборот)

По коду - equals для сравнения дат лучше не использовать, есть аналогичный метод isEqual

vlolad commented on src/manager/TaskMeneger.java:
```diff
@@ -48,4 +48,10 @@ public interface TaskMeneger {
+
+    List<PreTask> getPrioritizedTasks();
+
+    boolean isIntersection(PreTask preTask1);
```
Методы проверки пересечений должны быть приватные

vlolad reviewed (changes requested):
Привет!

Хорошая работа, направление работы по ТЗ верное и в целом все фичи новые реализованы. Пока не выполняется требование по оптимизации новых методов в менеджере, оставил в этой части комментарии, надо поправить работу с отсортированными по времени задачами и поиском пересечений

В качестве доп. пожелания - лучше форматирование задач в строки вынести в класс-конвертер, чтобы не смешивать этот функционал с менеджером

Буду ждать на повторную проверку, хорошего вечера!

AntonyukV516 committed:
Понимаю что программа не работает, но очень прошу помочь разобраться как это все исправить. При добавлении treeSet начали сыпаться тесты и я не понимаю как все наладить, тем более после очень долгого перерыва. Я не понимаю даже какие вопросы писать куратору и одногруппникам поэтому так отправляю.Спасибо

vlolad commented on src/manager/InMemoryTaskManager.java:
```diff
@@ -141,37 +151,43 @@ public Subtask getSubtaskById(Integer id) {
-            return oldEpic;
-        }
-        return null;
+        if (!isIntersection(updatedEpic)) {
```
В целом т.к. непосредственно у эпика не меняется время само по себе (только у подзадач), тут можно проверку не добавлять

vlolad commented on src/manager/InMemoryTaskManager.java:
```diff
@@ -222,4 +238,35 @@ public ArrayList<Subtask> getSubTaskFromEpic(Integer epicId) {
+            return false;
+        }
+        return !(startTime1.isBefore(startTime2) && endTime1.isBefore(endTime2))
+                || endTime1.equals(startTime2)
```
Лучше `isEqual`, метод как раз заточен под использование с датами и расширяет стандартную реализацию `equals`

vlolad commented on src/model/Epic.java:
```diff
@@ -28,4 +35,51 @@ public void removeSubtaskId(Integer id) {
+
+    public Instant getStartTime(TaskMeneger taskMeneger) {
+        try {
+            return taskMeneger.getSubTaskFromEpic(this.id)
```
Не очень хорошая практика вызывать менеджер из модели. Лучше поле startTime и duration в эпике просто обновлять, когда обновляется (или добавляется) его подзадача

vlolad reviewed (approved):
Привет!

Хорошая работа, с пересечениями проблем больше нет. Есть несколько минорных замечаний, но, думаю, их можно и до следующего спринта поправить. Главное, что всё работает)

Хорошего вечера!

AntonyukV516 committed:
Исправил замечания, переделал, закомментировал тесты,
 так как проблемы думаю в них, на этапе тестирования сервера уже будет понятно что работает корректно а что не очень

AntonyukV516 committed:
исправлен codeStyle
